src/catalog/src/durable/impls/persist.rs

```diff
@@ -453,6 +456,39 @@ impl PersistHandle {
             .sorted_by(|a, b| Ord::cmp(&b.ts, &a.ts))
     }
 
+    /// Generates an iterator of [`StateUpdate`] that contain all updates to the catalog
+    /// state up to, but not including, `upper`.
+    async fn snapshot_unconsolidated(
+        &mut self,
+        upper: Timestamp,
+    ) -> impl Iterator<Item = StateUpdate> + DoubleEndedIterator {
+        let snapshot = if upper > 0.into() {
+            let since = self.read_handle.since().clone();
+            let mut as_of = upper.saturating_sub(1);
+            as_of.advance_by(since.borrow());
+            let mut snapshot = Vec::new();
+            let mut stream = Box::pin(
+                self.read_handle
+                    .snapshot_and_stream(Antichain::from_elem(as_of))
+                    .await
+                    .expect("we have advanced the restart_as_of by the since"),
+            );
+            while let Some(update) = stream.next().await {
+                snapshot.push(update)
+            }
+            snapshot
+        } else {
+            Vec::new()
+        };
+        snapshot
+            .into_iter()
+            .map(|((key, _unit), ts, diff)| StateUpdate {
```

Comment:
nit: if you like, you can probably replace `_unit` with something like `&()`

Reply:
That results in the following compiler error:

```
error[E0308]: mismatched types
   --> src/catalog/src/durable/impls/persist.rs:485:26
    |
485 |             .map(|((key, &()), ts, diff)| StateUpdate {
    |                          ^^^           - expected due to this
    |                          |
    |                          expected `Result<(), String>`, found `&_`
    |
    = note:   expected enum `Result<(), std::string::String>`
            found reference `&_`
```

Reply:
Ah, sorry I wasn't looking closely and missed the Result. Feel free to ignore